### PR TITLE
Issue#2760 convert data source.java

### LINKED ISSUE #2760: Convert Datasource.java to Datasource.kt
parent issue #1826

What changes you will do:
will convert Java file to Kotlin

I couldn't find an issue regarding this, do tell me if there is I will close this

### PR DESCRIPTION
**Pull request by cdhiraj40**
fixes #2760

### CODE REVIEW

File: app/src/main/java/org/kiwix/kiwixmobile/zim_manager/ZimManageViewModel.kt [hunk 1]
```diff
@@ -465,8 +465,12 @@ class ZimManageViewModel @Inject constructor(
...
         { newList ->
           fileSelectListStates.postValue(
-            fileSelectListStates.value?.let { inheritSelections(it, newList) }
-              ?: FileSelectListState(newList)
+            fileSelectListStates.value?.let {
+              inheritSelections(
+                it,
+                newList as MutableList<BooksOnDiskListItem>
+              )
+            } ?: FileSelectListState(newList)
```
Comment (gouri-panda) [inline app/src/main/java/org/kiwix/kiwixmobile/zim_manager/ZimManageViewModel.kt:473]: Not sure, why do you want to implement this in this PR?

File: core/src/main/java/org/kiwix/kiwixmobile/core/data/DataSource.kt [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+  fun saveLanguages(languages: List<Language>): Completable?
+  fun saveHistory(history: HistoryItem): Completable?
+  fun deleteHistory(historyList: List<HistoryListItem>): Completable?
+  fun clearHistory(): Completable?
+  fun getBookmarks(): Flowable<List<BookmarkItem>>
+  fun getCurrentZimBookmarksUrl(): Single<List<String>>
+
+  fun saveBookmark(bookmark: BookmarkItem): Completable?
+  fun deleteBookmarks(bookmarks: List<BookmarkItem>): Completable?
+  fun deleteBookmark(bookmarkUrl: String): Completable?
```
Comment (gouri-panda) [inline core/src/main/java/org/kiwix/kiwixmobile/core/data/DataSource.kt:47]: Here `Completable` shouldn't be null. If something goes wrong for example database error, we can return `Failure`.

File: core/src/main/java/org/kiwix/kiwixmobile/core/main/MainRepositoryActions.kt [hunk 3]
```diff
@@ -31,16 +31,16 @@ class MainRepositoryActions @Inject constructor(private val dataSource: DataSour
 
   fun saveHistory(history: HistoryItem) {
     dataSource.saveHistory(history)
-      .subscribe({}, { e -> Log.e(TAG, "Unable to save history", e) })
+      ?.subscribe({}, { e -> Log.e(TAG, "Unable to save history", e) })
```
Comment (gouri-panda) [inline core/src/main/java/org/kiwix/kiwixmobile/core/main/MainRepositoryActions.kt:34]: In my previous comment case therefore we don't need to check null here.

File: README.md [hunk 4]
```diff
@@ -23,6 +23,7 @@ pieces in Java).
 [![License: GPL v3](https://img.shields.io/badge/License-GPLv3-blue.svg)](https://www.gnu.org/licenses/gpl-3.0)
 [![Public Chat](https://img.shields.io/badge/public-chat-green)](https://chat.kiwix.org)
 [![Slack](https://img.shields.io/badge/Slack-chat-E01E5A)](https://kiwixoffline.slack.com)
+[![Nightly](https://github.com/kiwix/kiwix-android/actions/workflows/nightly.yml/badge.svg)](https://github.com/kiwix/kiwix-android/actions/workflows/nightly.yml)
```
[the remark this answers is not in the document]
Comment (cdhiraj40) [reply README.md:26]: agreed, it's an old commit that is already merged. As I said on slack its coming everytime with the new PR :/

File: app/src/main/java/org/kiwix/kiwixmobile/zim_manager/ZimManageViewModel.kt [hunk 5]
```diff
@@ -270,7 +270,7 @@ class ZimManageViewModel @Inject constructor(
   ) = Flowable.combineLatest(
     booksFromDao,
     downloads,
-    languages.filter { it.isNotEmpty() },
+    languages.filter(List<Language>::isNotEmpty),
```
[the remark this answers is not in the document]
Comment (cdhiraj40) [reply app/src/main/java/org/kiwix/kiwixmobile/zim_manager/ZimManageViewModel.kt:273]: okay sorry for that will not change which is not related to PR.

File: app/src/main/java/org/kiwix/kiwixmobile/zim_manager/ZimManageViewModel.kt (hunk 1 above)
Comment (cdhiraj40) [reply app/src/main/java/org/kiwix/kiwixmobile/zim_manager/ZimManageViewModel.kt:473]: there was an error after changing the code.
```
 Type mismatch: inferred type is List<BooksOnDiskListItem>! but MutableList<BooksOnDiskListItem> was expected
```

File: core/src/main/java/org/kiwix/kiwixmobile/core/data/DataSource.kt (hunk 2 above)
Comment (cdhiraj40) [reply core/src/main/java/org/kiwix/kiwixmobile/core/data/DataSource.kt:47]: okay sure will change it :)

**Comment by cdhiraj40** [comment]
Closing this PR and opening a new one as there were problems with this branch, Sorry for the inconvenience.

**Pull request closed without merging**